offload/plugins-nextgen/common/include/PluginInterface.h

```
@@ -58,6 +58,30 @@ struct GenericKernelTy;
 struct GenericDeviceTy;
 struct RecordReplayTy;
 
+enum class ErrorCode {
+#define OFFLOAD_ERRC(Name, _, Value) Name = Value,
+#include "OffloadErrcodes.inc"
+#undef OFFLOAD_ERRC
+};
+
+class OffloadErrorCategory : public std::error_category {
+  const char *name() const noexcept override { return "Offload Error"; }
+  std::string message(int ev) const override {
+    switch (static_cast<ErrorCode>(ev)) {
+#define OFFLOAD_ERRC(Name, Desc, Value)                                        \
+  case ErrorCode::Name:                                                        \
+    return #Desc;
+#include "OffloadErrcodes.inc"
+#undef OFFLOAD_ERRC
+    }
+  }
+};
+
+inline std::error_code make_error_code(ErrorCode EC) {
+  static OffloadErrorCategory Cat{};
+  return {static_cast<int>(EC), Cat};
+}
```

What do we really need `std::error_code` for? I thought we were just going from LLVM error to error code + string pair in the API.

---

I'm just wondering if this is really necessary, going on comments like this which do something similar to what you're doing here https://github.com/llvm/llvm-project/blob/f49ee00ec47f596269639987f5ea8755a358ca16/llvm/lib/Object/Error.cpp#L21.

---

The createStringError message takes a std::error_code as a parameter, which is why I added all the error_code machinery.

I also think it makes sense to use here, since it also allows us to inspect whether a given error code is an "offload" error code or some other kind of error that has snuck in (which is what GetErrorCode does).

---

Yeah, but shouldn't we be able to use `https://en.cppreference.com/w/cpp/error/errc/make_error_code` for the generic category? I don't think having an `offload` category helps us here since it's not used for anything. The string message is going to come from the person who created the error, so we only care about the numerical part of the error code.

---

So to give a concrete example, as part of parsing elf files, we call `ELFObjectFileBase::createELFObjectFile` which can fail (e.g. if the elf file is malformed). When it fails, it returns some llvm::Error. Since this is not an offload error, we could/should swallow it and convert it to an appropriate Offload error (like INVALID_BINARY or whatever). If we forget to do that, we end up with the "Elf Error code" being returned from the C API, which is going to be nonsensical. That's why I think it's important to be able to tell whether an error is "offload" or not, so we can convert the nonsensical ELF errors into `UNKNOWN` or similar.

Although, now I've looked into it more, it looks like the preferred method is to use error RTTI for this. Would you be okay with it if, instead, I made an OffloadError type that extends StringError like some other components do? https://llvm.org/doxygen/classllvm_1_1StringError.html

---

That's a more compelling example, but I'm wondering if we shouldn't just convert it when we create the error rather than late. Doing it late requires disambiguating it, but if we just do it when we make it then it's pretty straightforward what it should be.

---

That's a little obtuse, but I guess if it's canonically obtuse it's okay.

---

The problem is StringError is built around std::error_code. So I guess once/if that's changed it should be easy to take out the std::error_code machinery.